fix: avoid selecting subnets with insufficient available IP address

Vacant2333 opened the pull request:
<!-- Please follow the guidelines at https://www.conventionalcommits.org/en/v1.0.0/ and use one of the following in your title:
feat:            <-- New features that require a MINOR version update
fix:             <-- Bug fixes that require at PATCH version update
chore:           <-- Smaller changes that impact behavior but aren't large enough to be features
perf:            <-- Code changes that improve performance but do not impact behavior
docs:            <-- Documentation changes that do not impact code
test:            <-- Test changes that do not impact behavior
ci:              <-- Changes that affect test or rollout automation
!${type}:        <-- Include ! if your change includes a backwards incompatible change.

Please review the Karpenter contribution docs at https://karpenter.sh/docs/contributing/ before submitting your pull request.
-->

Fixes https://github.com/aws/karpenter-provider-aws/issues/5234
https://github.com/aws/karpenter-provider-aws/issues/2921

**Description**
its a feature that when launchInstance select the subnet which has the most available ip count.

**How was this change tested?**

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates <!-- docs must be added to /preview to be included in future version releases -->
- [ ] Yes, issue opened: https://github.com/aws/karpenter-provider-aws/issues/5234
- [x] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

Vacant2333 commented:
@jmdeal @jonathan-innis hi, can u help me review, this pr will resolve the issue https://github.com/aws/karpenter-provider-aws/issues/5234

Talbalash-legit commented:
@rschalo I know you're incredibly busy, but this PR would be a game-changer for us. It addresses the persistent issue we've been struggling with in [#5234](https://github.com/aws/karpenter-provider-aws/issues/5234), and getting it merged would genuinely make a world of difference for our team. We're counting on your expertise to give it the green light. Thank you so much for considering this 🙏

rschalo commented:
Hi All - there is not currently a unit test around this behavior. Could we add one where we assert a subnet is no longer considered if IP exhaustion is expected? From a quick look I think it'd be tested via targeting one zone in a test for InstanceProvider create and an with update to the mock EC2 provider that adds a subnet with 0 available IPs for the targeted zone but not sure how that affects other tests.

Talbalash-legit commented:
Hey can we have an update on this please? 🙏  @rschalo

saurav-agarwalla commented:
/assign

saurav-agarwalla commented:
Can we bring in https://github.com/aws/karpenter-provider-aws/pull/7549 to this PR? Related changes should be a part of the same PR.

Can you please also address the errors flagged by the CI?

saurav-agarwalla commented on pkg/providers/subnet/subnet.go:
```diff
@@ -168,6 +168,13 @@ func (p *DefaultProvider) ZonalSubnetsForLaunch(ctx context.Context, nodeClass *
+
+		// Check if the remaining IP count is insufficient to meet the predicted IP usage;
+		// if so, remove this subnet zone record from inflightIPs and continue to the next item in the loop。
+		if prevIPs-predictedIPsUsed < 0 {
```
Shouldn't this be `<=` instead? I don't think we want to consider subnets with zero IPs.

Vacant2333 replied:
thanks for your review!

Vacant2333 commented:
> Can we bring in #7549 to this PR? Related changes should be a part of the same PR.
> 
> Can you please also address the errors flagged by the CI?

Hello @saurav-agarwalla ! I’m inviting the author of #7549  to co-edit this PR with me. I’ve already addressed your comments and fixed the CI issues. Thank you for helping me move this PR forward!
